### ISSUE TITLE
Doesn't say which snippet has a buggy definition

### ISSUE THREAD

**Issue by mgedmin** [SirVer/ultisnips#551]
I upgraded UltiSnips and tab-expansion stopped working.  I get the following error:
```
An error occured. This is either a bug in UltiSnips or a bug in a
snippet definition. If you think this is a bug, please report it to
https://github.com/SirVer/ultisnips/issues/new.

Following is the full stack trace:
Traceback (most recent call last):
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet_manager.py", line 59, in wrapper
    return func(self, *args, **kwds)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet_manager.py", line 152, in expand_or_jump
    rv = self._try_expand()
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet_manager.py", line 673, in _try_expand
    snippets = self._snips(before, False)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet_manager.py", line 555, in _snips
    source.ensure(filetypes)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/_base.py", line 43, in ensure
    self._load_snippets_for(ft)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/_base.py", line 74, in _load_snippets_for
    self._parse_snippets(ft, fn)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/_base.py", line 88, in _parse_snippets
    for event, data in self._parse_snippet_file(file_data, filename):
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/ultisnips.py", line 180, in _parse_snippet_file
    for event, data in _parse_snippets_file(filedata, filename):
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/ultisnips.py", line 147, in _parse_snippets_file
    actions
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/source/file/ultisnips.py", line 120, in _handle_snippet_or_global
    context, pre_expand)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/definition/_base.py", line 67, in __init__
    self.matches(self._trigger)
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/definition/_base.py", line 245, in matches
    self._context = self._context_match()
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/definition/_base.py", line 95, in _context_match
    'context': None
  File "/home/mg/.vim/bundle/UltiSnips/pythonx/UltiSnips/snippet/definition/_base.py", line 119, in _eval_code
    exec(code, {'snip': snip})
  File "<string>", line 16, in <module>
AttributeError: type object 'buffer' has no attribute 'name'
```

I assume one of my snippets no longer works, but the error doesn't tell me which one!  Grepping for 'name' in ~/.vim/UltiSnips/* gives me [this one](https://github.com/mgedmin/dotvim/blob/master/UltiSnips/rst.snippets#L34):

```
snippet travis "Travis CI badge" "github_owner_repo(buffer.name)" be
.. image:: https://travis-ci.org/`!p snip.rv = rv = github_owner_repo(path)`.svg?branch=master
    :target: https://travis-ci.org/`!p snip.rv = rv`

endsnippet
```

Commenting this snippet out confirms that it's the one that broke tab-completion (of an unrelated snippet "date", in a ReStructuredText file).

This bug is not about a new release of UltiSnips breaking a snippet that used to work.  It's about UltiSnips not telling me which snippet is causing the problem.

**Comment by SirVer** [comment]
Yes, that can be rather annoying.

**Comment by mirzalazuardi** [comment]
have you found the solution @mgedmin ??

**Comment by seletskiy** [comment]
@mgedmin, @SirVer: should be fixed in 698f2d49103f20c09bebad1656da26def81d36cf and following commit. After that commit ultisnips will show output like this in scratch buffer:

```
Snippet, caused error:

  Defined in: /home/s.seletskiy/.vim/UltiSnips/go.snippets:606
  Trigger: a
  Description: "Description"
  Context: asd()
  Pre-expand: <none>
  Post-expand: <none>

Executed snippet code:
   ...
```
